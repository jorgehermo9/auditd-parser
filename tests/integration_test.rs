use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
