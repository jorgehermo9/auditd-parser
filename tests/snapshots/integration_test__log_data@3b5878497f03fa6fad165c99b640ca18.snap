---
source: tests/integration_test.rs
expression: record
info: "type=EXECVE msg=audit(1491946471.575:1075898): argc=3 a0=\"jq\" a1=\".\" a2=7B0A202020202020202022646576223A202230383A3031222C0A202020202020202022696E6F6465223A20223139353439363436222C0A2020202020202020226974656D223A202230222C0A2020202020202020226D6F6465223A202230313030373735222C0A2020202020202020226E616D65223A20222F7573722F62696E2F6A71222C0A2020202020202020226F626A223A2022756E636F6E66696E65645F753A6F626A6563745F723A757365725F686F6D655F743A7330222C0A2020202020202020226F626A74797065223A20224E4F524D414C222C0A2020202020202020226F676964223A202231303031222C0A2020202020202020226F756964223A202231303030222C0A2020202020202020227261775F6D657373616765223A2022617564697428313439313934363239362E3735373A31303735383334293A206974656D3D30206E616D653D5C222F7573722F62696E2F6A715C2220696E6F64653D3139353439363436206465763D30383A3031206D6F64653D30313030373735206F7569643D31303030206F6769643D3130303120726465763D30303A3030206F626A3D756E636F6E66696E65645F753A6F626A6563745F723A757365725F686F6D655F743A7330206F626A747970653D4E4F524D414C222C0A20202020202020202272646576223A202230303A3030222C0A2020202020202020227265636F72645F74797065223A202250415448222C0A20202020202020202273657175656E6365223A20313037353833340A2020202020207D"
---
{
  "record_type": "EXECVE",
  "timestamp": 1491946471575,
  "id": 1075898,
  "fields": {
    "a0": "jq",
    "a1": ".",
    "a2": "7B0A202020202020202022646576223A202230383A3031222C0A202020202020202022696E6F6465223A20223139353439363436222C0A2020202020202020226974656D223A202230222C0A2020202020202020226D6F6465223A202230313030373735222C0A2020202020202020226E616D65223A20222F7573722F62696E2F6A71222C0A2020202020202020226F626A223A2022756E636F6E66696E65645F753A6F626A6563745F723A757365725F686F6D655F743A7330222C0A2020202020202020226F626A74797065223A20224E4F524D414C222C0A2020202020202020226F676964223A202231303031222C0A2020202020202020226F756964223A202231303030222C0A2020202020202020227261775F6D657373616765223A2022617564697428313439313934363239362E3735373A31303735383334293A206974656D3D30206E616D653D5C222F7573722F62696E2F6A715C2220696E6F64653D3139353439363436206465763D30383A3031206D6F64653D30313030373735206F7569643D31303030206F6769643D3130303120726465763D30303A3030206F626A3D756E636F6E66696E65645F753A6F626A6563745F723A757365725F686F6D655F743A7330206F626A747970653D4E4F524D414C222C0A20202020202020202272646576223A202230303A3030222C0A2020202020202020227265636F72645F74797065223A202250415448222C0A20202020202020202273657175656E6365223A20313037353833340A2020202020207D",
    "argc": "3"
  }
}
